FF-2022 Add bandits

rasendubi opened the pull request.

rasendubi commented on eppoclient/assignmentlogger.go:
```diff
@@ -6,6 +6,12 @@ type IAssignmentLogger interface {
 	LogAssignment(event AssignmentEvent)
 }
 
+// BanditActionLogger is going to be merged into IAssignmentLogger in
+// the next major version.
+type BanditActionLogger interface {
+	LogBanditAction(event BanditEvent)
+}
```
This is to limit upgrade to a minor version bump. If we add `LogBanditAction` to `IAssignmentLogger` that would require a major bump.

See https://go.dev/blog/module-compatibility > working with interfaces

rasendubi commented on eppoclient/evalbandits.go:
```diff
@@ -0,0 +1,235 @@
+// - Integer and float types are mapped to numeric attributes.
+// - Strings and bools are mapped to categorical attributes.
+// - Rest of types are silently dropped.
+func InferContextAttributes(attrs map[string]interface{}) ContextAttributes {
```
Only exposed for user's convenience

rasendubi commented on eppoclient/client.go:
```diff
@@ -80,6 +81,76 @@ func (ec *EppoClient) GetJSONAssignment(flagKey string, subjectKey string, subje
+		}
+	}
+
+	evaluation := bandit.ModelData.evaluate(banditEvaluationContext{
+		flagKey:           flagKey,
+		subjectKey:        subjectKey,
+		subjectAttributes: subjectAttributes,
+		actions:           actions,
+	})
+	if evaluation.result == nil {
+		return BanditResult{
+			Variation: variation,
+			Action:    nil,
+		}
+	}
+
+	if logger, ok := ec.logger.(BanditActionLogger); ok {
```
One difference from Python SDK is that empty evaluations are not logged. Empty evaluations can only happen if `actions` passed is empty, so I'm not sure how helpful that is.

It's a bit cumbersome to work with nullable values in Go, so I decided to not log them. I hope that's fine. But if not—let me know and I'll rework it.

cc @schmit

aarsilv replied:
Across the board, for bandits and non-bandit flags, if our evaluation hit an error and we're going to return the default, we do not want to log as is its not an "intentional" assignment.

I think if empty actions are passed, we'd want to return (and not log) the default value.

aarsilv replied:
Moved this discussion to Slack here: https://eppo-group.slack.com/archives/C05R6BK5BB9/p1718906479473079

aarsilv commented on eppoclient/bandits_test.go:
```diff
@@ -0,0 +1,134 @@
+	}
+}
+
+func Test_InferContextAttributes(t *testing.T) {
```
🙌

aarsilv commented on eppoclient/bandits_test.go:
```diff
@@ -0,0 +1,134 @@
+	assert.Equal(t, expected, contextAttributes)
+}
+
+func Test_bandits_sdkTestData(t *testing.T) {
```
🔥

aarsilv commented on eppoclient/bandits_test.go:
```diff
@@ -0,0 +1,134 @@
+	}
+	logger := new(mockLogger)
+	logger.Mock.On("LogAssignment", mock.Anything).Return()
+	logger.Mock.On("LogBanditAction", mock.Anything).Return()
```
we should have _one_ test and test case where we spy on `LogBanditAction` and make sure the right things are being passed as the event.

rasendubi replied:
I have added a new test for this

aarsilv commented on eppoclient/bandits_test.go:
```diff
@@ -0,0 +1,134 @@
+	}
+}
+
+func readJsonFile[T any](filePath string) T {
```
This and other processing and serving test data files we already need to do for [`eppoclient_e2e_test.go`](https://github.com/Eppo-exp/golang-sdk/blob/main/eppoclient/eppoclient_e2e_test.go#L92), so consider consolidating/deduplicating the code into some sort of reusable helper function.

aarsilv commented on eppoclient/client.go:
```diff
@@ -2,26 +2,27 @@ package eppoclient
-	ec.logger = assignmentLogger
-
-	return ec
+func newEppoClient(configurationStore *configurationStore, configRequestor *configurationRequestor, poller *poller, assignmentLogger IAssignmentLogger) *EppoClient {
```
ah so this is not a breaking change because it's used by [`initclient.go`](https://github.dev/Eppo-exp/golang-sdk/blob/main/eppoclient/eppoclient_e2e_test.go), which is the entry point for our API. 👌

aarsilv commented on eppoclient/client.go:
```diff
@@ -80,6 +81,76 @@ func (ec *EppoClient) GetJSONAssignment(flagKey string, subjectKey string, subje
+	variation, _ := ec.GetStringAssignment(flagKey, subjectKey, subjectAttributes.toGenericAttributes(), defaultVariation)
+
+	bandit, err := ec.configurationStore.getBanditConfiguration(variation)
+	if err != nil {
```
maybe this is a normal go thing, but it seems odd to use an error to signify there is no bandit, as opposed to just returning a `nil` bandit.

aarsilv commented on eppoclient/client.go:
```diff
@@ -80,6 +81,76 @@ func (ec *EppoClient) GetJSONAssignment(flagKey string, subjectKey string, subje
+		subjectAttributes: subjectAttributes,
+		actions:           actions,
+	})
+	if evaluation.result == nil {
```
This seems like an error situation--at this point should we instead return the default value?

rasendubi replied:
`BanditResult{variation, nil}` is the default value in this case. This function does not receive a default action to return (only default variation).

This is the same interface/behavior as we have in Python. cc @schmit

aarsilv commented on eppoclient/client.go:
```diff
@@ -80,6 +81,76 @@ func (ec *EppoClient) GetJSONAssignment(flagKey string, subjectKey string, subje
+			ActionProbability:            evaluation.result.actionWeight,
+			OptimalityGap:                evaluation.result.optimalityGap,
+			ModelVersion:                 bandit.ModelVersion,
+			Timestamp:                    time.Now().UTC().Format(time.RFC3339),
```
Just want to call out we're getting slightly consistent with our timestamp logging precission. 
* This will log to the second
* [Python SDK]("timestamp": datetime.datetime.utcnow().isoformat(),) logs the microsecond 
* [Java SDK](https://github.com/Eppo-exp/sdk-common-jdk/blob/main/src/main/java/cloud/eppo/rac/dto/BanditLogData.java#L31) keeps things as a `Date` object, which is precise to the millisecond.

I actually like the Java approach the most: keeping the date in a native datetime format and letting their logger implementation handle it. So in this case we'd just pass `time.Now().UTC()` and omit the formatting part.

rasendubi replied:
I would like to keep the `Timestamp` the same type between `AssignmentEvent` and `BanditEvent`. I'll leave a todo that we should update the type in the next major release.

aarsilv replied:
Ah yes agree that consistency would be best

aarsilv commented on eppoclient/client_test.go:
```diff
@@ -132,13 +120,41 @@ func Test_GetStringAssignmentHandlesLoggingPanic(t *testing.T) {
 	assert.Equal(t, expected, assignment)
 }
+
+func Test_client_handlesBanditLoggerPanic(t *testing.T) {
```
nice!

aarsilv commented on eppoclient/client_test.go:
```diff
@@ -132,13 +120,41 @@ func Test_GetStringAssignmentHandlesLoggingPanic(t *testing.T) {
+func Test_client_handlesBanditLoggerPanic(t *testing.T) {
+	var logger = new(mockLogger)
+	logger.Mock.On("LogAssignment", mock.Anything).Return()
+	logger.Mock.On("LogBanditAction", mock.Anything).Panic("logging panic")
```
🔥

aarsilv commented on eppoclient/client_test.go:
```diff
@@ -132,13 +120,41 @@ func Test_GetStringAssignmentHandlesLoggingPanic(t *testing.T) {
+	actions := map[string]ContextAttributes{
+		"action1": {},
+	}
+	client.GetBanditAction("testFlag", "subject", ContextAttributes{}, actions, "bandit")
```
we should probably assert what is returned here, making sure it's _not_ the default value.

rasendubi replied:
I have added another test case that asserts that.

aarsilv commented on eppoclient/configurationrequestor_test.go:
```diff
@@ -0,0 +1,62 @@
+	"github.com/stretchr/testify/assert"
+)
+
+func Test_configurationRequestor_requestBandits(t *testing.T) {
```
🙌

aarsilv commented on eppoclient/configurationrequestor_test.go:
```diff
@@ -0,0 +1,62 @@
+	assert.NotEmpty(t, configurationStore.bandits)
+}
+
+func Test_configurationRequestor_shouldNotRequestBanditsIfNotPresentInFlags(t *testing.T) {
```
🙌

aarsilv commented on eppoclient/evalbandits.go:
```diff
@@ -0,0 +1,234 @@
+	}
+	for key, value := range attrs {
+		switch value := value.(type) {
+		case int:
+			result.Numeric[key] = float64(value)
+		case int8:
+			result.Numeric[key] = float64(value)
+		case int16:
+			result.Numeric[key] = float64(value)
+		case int32:
+			result.Numeric[key] = float64(value)
+		case int64:
+			result.Numeric[key] = float64(value)
+		case uint:
+			result.Numeric[key] = float64(value)
+		case uint8:
+			result.Numeric[key] = float64(value)
+		case uint16:
+			result.Numeric[key] = float64(value)
+		case uint32:
+			result.Numeric[key] = float64(value)
+		case uint64:
+			result.Numeric[key] = float64(value)
+		case float32:
+			result.Numeric[key] = float64(value)
+		case float64:
+			result.Numeric[key] = value
+		case string:
+			result.Categorical[key] = value
+		case bool:
+			result.Categorical[key] = strconv.FormatBool(value)
+		}
```
We're a lof of the same thing here. Would it be possible to group everything together so we just have one call to `float64(value)`?

rasendubi replied:
Unfortunately, no. When calling `float64()`, Go has to know the type of argument and the only way to do that is to single out every possible option

aarsilv commented on eppoclient/evalbandits.go:
```diff
@@ -0,0 +1,234 @@
+}
+
+func (model *banditModelData) evaluate(ctx banditEvaluationContext) banditEvaluationDetails {
+	// THere's currently no way to change totalShards in bandit evaluation.
```
minor but you got one extra capital letter 😛

aarsilv commented on eppoclient/evalbandits.go:
```diff
@@ -0,0 +1,234 @@
+	var totalShards int64 = 10_000
+
+	nActions := len(ctx.actions)
+	if nActions == 0 {
+		return banditEvaluationDetails{
+			flagKey:           ctx.flagKey,
+			subjectKey:        ctx.subjectKey,
+			subjectAttributes: ctx.subjectAttributes,
+			result:            nil,
+		}
+	}
```
per my other comment, I wonder if we should prevent getting here upstream and just return default value. That way the top-level function always returns something usable.

rasendubi replied:
I have updated the upstream function. Now `banditEvaluationDetails.result` can never be `nil`, so I inlined it.

aarsilv commented on eppoclient/evalbandits.go:
```diff
@@ -0,0 +1,234 @@
+	weights[bestAction] = remainderWeight
+
+	actions := make([]string, nActions)
+	shards := make(map[string]int64)
```
Consider a comment explaining what we are doing here: deterministically randomly shuffling the shards, so that as weights grow and shrink a huge tranch of users is not shifted from one old action to all now getting the same new action

aarsilv commented on eppoclient/evalbandits.go:
```diff
@@ -0,0 +1,234 @@
+	remainderWeight := math.Max(0.0, 1.0-totalWeight)
+	weights[bestAction] = remainderWeight
+
+	actions := make([]string, nActions)
```
since the evaluation context also has actions, consider naming this something like `actionKeys`

rasendubi replied:
renamed to `shuffledActions`

aarsilv commented on eppoclient/evalbandits.go:
```diff
@@ -0,0 +1,234 @@
+	weights[bestAction] = remainderWeight
+
+	actions := make([]string, nActions)
+	shards := make(map[string]int64)
```
Also, consider naming this something like `actionShards`

rasendubi replied:
I have scoped this variable to a smaller block, so the name should be unambiguous now.

aarsilv commented on eppoclient/evalbandits.go:
```diff
@@ -0,0 +1,234 @@
+		a2 := actions[j]
+		v1 := shards[a1]
+		v2 := shards[a2]
+		if v1 < v2 {
+			return true
+		} else if v1 > v2 {
+			return false
+		} else {
+			// tie-breaking
+			return a1 < a2
```
I think comments could go a long way here. We're "shuffling" the actions by sorting by a randomly assigned shard number (which--while unlikely--two actions _may_ share), and then tie breaking with action names, which two actions will not share.

aarsilv commented on eppoclient/evalbandits.go:
```diff
@@ -0,0 +1,234 @@
+	})
+
+	shard := getShard(ctx.flagKey+"-"+ctx.subjectKey, totalShards)
+	shardValue := float64(shard) / float64(totalShards)
```
I'm not sure this is the best name for what this variable represents. But I can't think of a better name 😛 

Maybe `actionAssignmentWeight`?

aarsilv commented on eppoclient/evalbandits.go:
```diff
@@ -0,0 +1,234 @@
+	var selectedAction string
+	for _, selectedAction = range actions {
+		cumulativeWeight += weights[selectedAction]
+		if cumulativeWeight > shardValue {
+			break
```
This is a neat way to do this. I _think_ we want `>=` though.

rasendubi replied:
`>` vs `>=` needs to be coordinated with Python implementation.

I don't know which one makes more sense. And it should probably not matter as we're dealing with fractional floats, so equality is somewhat unlikely anyway—dividing by 10,000 (totalShards) messes up numbers.

cc @schmit

aarsilv replied:
Thinking about this more and working through some cartoon examples, `>` is good 👍

aarsilv commented on eppoclient/evalbandits.go:
```diff
@@ -0,0 +1,234 @@
+		}
+	}
+
+	optimalityGap := bestScore - scores[selectedAction]
```
🔥

aarsilv reviewed (approved):
👏 GREAT stuff! 🎉 

Math and business logic look sound. Evaluation result looks good aside from us deciding as a team if we want string or native datetime timestamps (my vote is the latter). Also, given the complexity of the code consider layering in more comments and/or using functions to make it more clear what each step is doing.

I left a bunch of comments you are welcome to factor in or ignore as you see fit!

giorgiomartini0 reviewed (approved):
Nothing to add to @aarsilv 's very thorough review. Math and bandit logic look good to me.

rasendubi commented on eppoclient/banditmodel_test.go:
```diff
@@ -0,0 +1,27 @@
+		t.Fatal(err)
+	}
+
+	assert.NotEmpty(t, banditResponse.Bandits)
```
I would rather ignore it here—don't want the tests to know any content from `sdk-test-data`. We have `sdkTestData` tests that test that evaluation is performed correctly, which covers parsing part as well.

rasendubi commented:
I have refactored the bandit evaluation function and added support for non-bandit flag test case. @aarsilv @giorgiomartini0 would be great if you could take another look

giorgiomartini0 commented on eppoclient/bandits_test.go:
```diff
@@ -51,16 +51,16 @@ func Test_InferContextAttributes(t *testing.T) {
-		bandits: bandits.Bandits,
-	}
+	configStore := newConfigurationStore(configuration{
+		ufc:     ufc,
```
I feel like this should still be `flags` and `bandits`? Or bandit flags and bandit parameters. To me `ufc = flags UNION bandits`.

rasendubi replied:
Sounds good. I was confused because UFC is universal Flag configuration.

With this input, I think it doesn't make sense to use "ufc" in SDK at all as it's all UFC. I have pushed another commit renaming ufc to flags/config throughout the code.

giorgiomartini0 reviewed (commented):
Out of my depth and have not been following the latest context needed to give a full review. Nothing I saw looks off but please wait for others if you need a second round of approval.

rasendubi commented:
CI is failing depending on https://github.com/Eppo-exp/sdk-test-data/pull/39
